storage/src/main/java/org/apache/kafka/server/purgatory/DelayedRemoteFetch.java

```diff
@@ -0,0 +1,202 @@
+            return forceComplete();
+        }
+        return false;
+    }
+
+    @Override
+    public void onExpiration() {
+        // cancel the remote storage read task, if it has not been executed yet and
+        // avoid interrupting the task if it is already running as it may force closing opened/cached resources as transaction index.
+        remoteFetchTasks.forEach(((topicIdPartition, task) -> {
+            if (task != null && !task.isDone()) {
+                if (!task.cancel(false)) {
+                    LOG.debug("Remote fetch task for remoteFetchInfo: {} could not be cancelled.", remoteFetchInfos.get(topicIdPartition));
+                }
+            }
+        }));
+
+        EXPIRED_REQUEST_METER.mark();
+    }
+
+    /**
+     * Upon completion, read whatever data is available and pass to the complete callback
+     */
+    @Override
+    public void onComplete() {
+        Map<TopicIdPartition, FetchPartitionData> fetchPartitionData = new HashMap<>();
+        try {
+            for (Map.Entry<TopicIdPartition, LogReadResult> entry : localReadResults.entrySet()) {
+                TopicIdPartition tp = entry.getKey();
+                LogReadResult result = entry.getValue();
+
+                CompletableFuture<RemoteLogReadResult> remoteFetchResult = remoteFetchResults.get(tp);
+                if (remoteFetchInfos.containsKey(tp)
+                    && remoteFetchResult.isDone()
+                    && result.error() == Errors.NONE
+                    && result.info().delayedRemoteStorageFetch.isPresent()) {
+
+                    if (remoteFetchResult.get().error().isPresent()) {
+                        fetchPartitionData.put(tp,
+                            new LogReadResult(remoteFetchResult.get().error().get()).toFetchPartitionData(false));
+                    } else {
+                        FetchDataInfo info = remoteFetchResult.get().fetchDataInfo().get();
+                        fetchPartitionData.put(tp,
+                            new FetchPartitionData(
+                                result.error(),
+                                result.highWatermark(),
+                                result.leaderLogStartOffset(),
+                                info.records,
+                                Optional.empty(),
+                                result.lastStableOffset(),
+                                info.abortedTransactions,
+                                result.preferredReadReplica(),
+                                false));
+                    }
+                } else {
+                    fetchPartitionData.put(tp, result.toFetchPartitionData(false));
+                }
+            }
+        } catch (InterruptedException | ExecutionException e) {
+            throw new RuntimeException(e);
+        }
+
+        responseCallback.accept(fetchPartitionData);
+    }
+
+    // Visible for testing
+    public static Meter expiredRequestMeter() {
```

Comment:
the tests only need the `count`. Could you expose `EXPIRED_REQUEST_METER.count` rather than the full object